### PR TITLE
[VarDumper] Fix FFI caster test

### LINKED ISSUE #57387: FFICasterTest::testCastNonTrailingCharPointer relies on undefined behaviors
### Symfony version(s) affected

7.2

### Description

I believe that the test [FFICasterTest::testCastNonTrailingCharPointer
](https://github.com/symfony/symfony/blob/678abb4b128c0bd9f6db83a280bd5355ce234aec/src/Symfony/Component/VarDumper/Tests/Caster/FFICasterTest.php#L205) relies on undefined behavior, and will break depending on what is allocated just after `\FFI::cdef()->new('char['.$actualLength.']')`.

The assignment `$pointer[$actualLength] = "\x01";` writes 1 byte just _after_ the memory block allocated by `\FFI::cdef()->new('char['.$actualLength.']')`. Valgrind detects this invalid write when running with `USE_ZEND_ALLOC=0 valgrind ./phpunit ...`:

```
==460210== Invalid write of size 1
==460210==    at 0x809ACB: zend_ffi_zval_to_cdata (ffi.c:808)
==460210==    by 0x809ACB: zend_ffi_cdata_write_dim (ffi.c:1476)
==460210==    by 0xDC899D: zend_assign_to_object_dim (zend_execute.c:1532)
==460210==    by 0xEA1E40: ZEND_ASSIGN_DIM_SPEC_CV_CONST_OP_DATA_CONST_HANDLER (zend_vm_execute.h:42809)
==460210==    by 0xEEDDBE: execute_ex (zend_vm_execute.h:61830)
==460210==    by 0xEF10E9: zend_execute (zend_vm_execute.h:62776)
==460210==    by 0xD7966D: zend_execute_script (zend.c:1906)
==460210==    by 0xCA4799: php_execute_script_ex (main.c:2516)
==460210==    by 0xCA48BC: php_execute_script (main.c:2556)
==460210==    by 0x10132F6: do_cli (php_cli.c:956)
==460210==    by 0x10142CD: main (php_cli.c:1330)
==460210==  Address 0x1e30ddec is 0 bytes after a block of size 12 alloc'd
==460210==    at 0x484282F: malloc (vg_replace_malloc.c:446)
==460210==    by 0xD1FFA1: __zend_malloc (zend_alloc.c:3268)
==460210==    by 0xD1BF43: _emalloc (zend_alloc.c:2765)
==460210==    by 0x816E73: zim_FFI_new (ffi.c:3862)
==460210==    by 0xED7A69: ZEND_DO_FCALL_SPEC_RETVAL_USED_HANDLER (zend_vm_execute.h:1988)
==460210==    by 0xED7A69: execute_ex (zend_vm_execute.h:57414)
==460210==    by 0xEF10E9: zend_execute (zend_vm_execute.h:62776)
==460210==    by 0xD7966D: zend_execute_script (zend.c:1906)
==460210==    by 0xCA4799: php_execute_script_ex (main.c:2516)
==460210==    by 0xCA48BC: php_execute_script (main.c:2556)
==460210==    by 0x10132F6: do_cli (php_cli.c:956)
==460210==    by 0x10142CD: main (php_cli.c:1330)
```

Additionally, [FFICaster::castFFIStringValue()](https://github.com/symfony/symfony/blob/678abb4b128c0bd9f6db83a280bd5355ce234aec/src/Symfony/Component/VarDumper/Caster/FFICaster.php#L115) treats `char*` pointers a NUL-terminated C strings, but there is no guarantee that a NUL byte will be found near that address. As a result, in the test, the function will dump any data that happens to exist after `$string`. In theory this may also crash if this causes the function to reach an unreadable memory region.

### How to reproduce

```
php ./phpunit src/Symfony/Component/VarDumper/Tests/Caster/FFICasterTest.php  --filter=testCastNonTrailingCharPointer
```

Commit 25360ef24951f1c6b83f8bf85fbdcaff4a1a40e1 in php-src changes the memory layout slightly, which causes the test output to change:

```
1) Symfony\Component\VarDumper\Tests\Caster\FFICasterTest::testCastNonTrailingCharPointer
Failed asserting that string matches format description.
--- Expected
+++ Actual
@@ @@
 FFI\CData<char*> size 8 align 8 {
-  cdata: "Hello World!%s"
+  cdata: b"Hello World!\x011RK`\x1EÐþ\x16\x7F\x00"
 }

/home/runner/work/symfony/symfony/src/Symfony/Component/VarDumper/Test/VarDumperTestTrait.php:52
/home/runner/work/symfony/symfony/src/Symfony/Component/VarDumper/Tests/Caster/FFICasterTest.php:219
```

(https://github.com/symfony/symfony/actions/runs/9497428991/job/26173977979)

Running any php version under valgrind reports an invalid write and read of size 1:

```
USE_ZEND_ALLOC=0 valgrind php ./phpunit src/Symfony/Component/VarDumper/Tests/Caster/FFICasterTest.php  --filter=testCastNonTrailingCharPointer
```

```
==462378== Invalid write of size 1
==462378==    at 0x138452F6: zend_ffi_zval_to_cdata (ffi.c:808)
==462378==    by 0x138452F6: zend_ffi_cdata_write_dim (ffi.c:1476)
==462378==    by 0x555FF2: zend_assign_to_object_dim (zend_execute.c:1534)
==462378==    by 0x5986EA: ZEND_ASSIGN_DIM_SPEC_CV_CV_OP_DATA_CONST_HANDLER (zend_vm_execute.h:51907)
==462378==    by 0x5A45F9: execute_ex (zend_vm_execute.h:57332)
==462378==    by 0x3458AF: ZEND_DO_FCALL_SPEC_OBSERVER_HANDLER (zend_vm_execute.h:2052)
==462378==    by 0x346A26: execute_ex.cold (zend_vm_execute.h:57256)
==462378==    by 0x3458AF: ZEND_DO_FCALL_SPEC_OBSERVER_HANDLER (zend_vm_execute.h:2052)
==462378==    by 0x346A26: execute_ex.cold (zend_vm_execute.h:57256)
==462378==    by 0x3458AF: ZEND_DO_FCALL_SPEC_OBSERVER_HANDLER (zend_vm_execute.h:2052)
==462378==    by 0x346A26: execute_ex.cold (zend_vm_execute.h:57256)
==462378==    by 0x3458AF: ZEND_DO_FCALL_SPEC_OBSERVER_HANDLER (zend_vm_execute.h:2052)
==462378==    by 0x346A26: execute_ex.cold (zend_vm_execute.h:57256)
==462378==  Address 0x1a4bbf4c is 0 bytes after a block of size 12 alloc'd
==462378==    at 0x484282F: malloc (vg_replace_malloc.c:446)
==462378==    by 0x5033C4: __zend_malloc (zend_alloc.c:3130)
==462378==    by 0x13842595: zim_FFI_new (ffi.c:3861)
==462378==    by 0x34594D: ZEND_DO_FCALL_SPEC_OBSERVER_HANDLER (zend_vm_execute.h:2086)
==462378==    by 0x346A26: execute_ex.cold (zend_vm_execute.h:57256)
==462378==    by 0x3458AF: ZEND_DO_FCALL_SPEC_OBSERVER_HANDLER (zend_vm_execute.h:2052)
==462378==    by 0x346A26: execute_ex.cold (zend_vm_execute.h:57256)
==462378==    by 0x3458AF: ZEND_DO_FCALL_SPEC_OBSERVER_HANDLER (zend_vm_execute.h:2052)
==462378==    by 0x346A26: execute_ex.cold (zend_vm_execute.h:57256)
==462378==    by 0x3458AF: ZEND_DO_FCALL_SPEC_OBSERVER_HANDLER (zend_vm_execute.h:2052)
==462378==    by 0x346A26: execute_ex.cold (zend_vm_execute.h:57256)
==462378==    by 0x3458AF: ZEND_DO_FCALL_SPEC_OBSERVER_HANDLER (zend_vm_execute.h:2052)
==462378== 
==462378== Invalid read of size 1
==462378==    at 0x1383BB40: zend_ffi_cdata_to_zval (ffi.c:590)
==462378==    by 0x1383BB40: zend_ffi_cdata_read_dim (ffi.c:1423)
==462378==    by 0x557BDD: UnknownInlinedFun (zend_execute.c:2845)
==462378==    by 0x557BDD: zend_fetch_dimension_address_read_R_slow (zend_execute.c:2884)
==462378==    by 0x582B29: ZEND_FETCH_DIM_R_SPEC_CV_CV_HANDLER (zend_vm_execute.h:50876)
==462378==    by 0x5A6B29: execute_ex (zend_vm_execute.h:61391)
==462378==    by 0x3458AF: ZEND_DO_FCALL_SPEC_OBSERVER_HANDLER (zend_vm_execute.h:2052)
==462378==    by 0x346A26: execute_ex.cold (zend_vm_execute.h:57256)
==462378==    by 0x3458AF: ZEND_DO_FCALL_SPEC_OBSERVER_HANDLER (zend_vm_execute.h:2052)
==462378==    by 0x346A26: execute_ex.cold (zend_vm_execute.h:57256)
==462378==    by 0x3458AF: ZEND_DO_FCALL_SPEC_OBSERVER_HANDLER (zend_vm_execute.h:2052)
==462378==    by 0x346A26: execute_ex.cold (zend_vm_execute.h:57256)
==462378==    by 0x3458AF: ZEND_DO_FCALL_SPEC_OBSERVER_HANDLER (zend_vm_execute.h:2052)
==462378==    by 0x346A26: execute_ex.cold (zend_vm_execute.h:57256)
==462378==  Address 0x1a4bbf4c is 0 bytes after a block of size 12 alloc'd
==462378==    at 0x484282F: malloc (vg_replace_malloc.c:446)
==462378==    by 0x5033C4: __zend_malloc (zend_alloc.c:3130)
==462378==    by 0x13842595: zim_FFI_new (ffi.c:3861)
==462378==    by 0x34594D: ZEND_DO_FCALL_SPEC_OBSERVER_HANDLER (zend_vm_execute.h:2086)
==462378==    by 0x346A26: execute_ex.cold (zend_vm_execute.h:57256)
==462378==    by 0x3458AF: ZEND_DO_FCALL_SPEC_OBSERVER_HANDLER (zend_vm_execute.h:2052)
==462378==    by 0x346A26: execute_ex.cold (zend_vm_execute.h:57256)
==462378==    by 0x3458AF: ZEND_DO_FCALL_SPEC_OBSERVER_HANDLER (zend_vm_execute.h:2052)
==462378==    by 0x346A26: execute_ex.cold (zend_vm_execute.h:57256)
==462378==    by 0x3458AF: ZEND_DO_FCALL_SPEC_OBSERVER_HANDLER (zend_vm_execute.h:2052)
==462378==    by 0x346A26: execute_ex.cold (zend_vm_execute.h:57256)
==462378==    by 0x3458AF: ZEND_DO_FCALL_SPEC_OBSERVER_HANDLER (zend_vm_execute.h:2052)
==462378== 
```

### Possible Solution

Allocate more memory to make room for the `$pointer[$actualLength] = "\x01";`. However, this assignment is not strictly necessary as the null byte after `$actualMessage` is not copied by `\FFI::memcpy($pointer, $actualMessage, $actualLength)`.

There is no way to prevent `FFICaster::castFFIStringValue()` from dumping unrelated data, but if that's ok, it may be possible to prevent it from reaching unreadable memory and crashing (providing that `char*` points to a readable memory address) by not stepping over page boundaries. E.g. don't dump past `$addr | (4096-1)`.

It may be possible to find the size of the memory block if its allocated by zend_mm, and if `$addr` points to the beginning of the block, by using the zend_mm API. E.g. use [`is_zend_mm() && is_zend_ptr($addr)`](https://github.com/php/php-src/blob/8fd1388a7f4bec803e6ac646a3e09d64cdbbf342/Zend/zend_alloc.h#L224) to find if `$addr` points to the beginning of a block allocated by zend_mm, and [`_zend_mm_block_size()`](https://github.com/php/php-src/blob/8fd1388a7f4bec803e6ac646a3e09d64cdbbf342/Zend/zend_alloc.h#L252) to find its size. However I do not necessarily recommend this.

### Additional Context

The issue was detected in https://github.com/symfony/symfony/pull/57379, and reported by @alexandre-daubois in https://github.com/php/php-src/pull/14546#issuecomment-2165190339

### PR DESCRIPTION
**Pull request by alexandre-daubois**
| Q             | A
| ------------- | ---
| Branch?       | 6.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Issues        | Fix #57387
| License       | MIT

Fixes the invalid memory writing thanks to Arnaud's hints. Not easy to understand exactly what's the point of `testCastNonTrailingCharPointer`, so I hope the test is still valid like this.

### CODE REVIEW

**Comment by alexandre-daubois** [comment]
cc @xabbuh this fixes CI on 8.4. I don't know if you've seen it, but just in case so you don't pull your hair out trying to figure out failure 😄

**Comment by xabbuh** [comment]
@alexandre-daubois 👍 thanks, I have seen it but I have no idea around this code. That's why I let @nicolas-grekas do the review here. :)

**Comment by nicolas-grekas** [comment]
I don't have the knowledge to review this PR.
Maybe @arnaud-lb can help?

File: src/Symfony/Component/VarDumper/Tests/Caster/FFICasterTest.php [hunk 1]
```diff
@@ -191,25 +190,13 @@ public function testCastCuttedPointerToChar()
...
         $actualMessage = 'Hello World!';
         $actualLength = \strlen($actualMessage);
 
-        $string = \FFI::cdef()->new('char['.$actualLength.']');
+        $string = \FFI::cdef()->new('char['.($actualLength + 1).']');
         $pointer = \FFI::addr($string[0]);
-
         \FFI::memcpy($pointer, $actualMessage, $actualLength);
 
         // Remove automatically addition of the trailing "\0" and remove trailing "\0"
```
Comment (arnaud-lb) [inline src/Symfony/Component/VarDumper/Tests/Caster/FFICasterTest.php]: This comment can be removed as it's incorrect

File: src/Symfony/Component/VarDumper/Caster/FFICaster.php [hunk 2]
```diff
@@ -115,12 +115,20 @@ private static function castFFIPointer(Stub $stub, CType $type, ?CData $data = n
...
+            if ($ffi->is_zend_ptr($data) && $ffi->is_zend_mm()) {
+                $result[$i] = $data[$i];
 
-            if ("\0" === $result[$i]) {
-                return implode('', $result);
+                if ("\0" === $data[$i]) {
+                    return implode('', $result);
+                }
+            } else {
+                break;
```
Comment (arnaud-lb) [inline src/Symfony/Component/VarDumper/Caster/FFICaster.php]: While testing I found that `is_zend_ptr()` is broken and may crash in some conditions. So my suggestion of using `is_zend_ptr()` is not good, sorry.

A safe alternative is to ensure that you only access memory in the same page as $data[0]: (it's safe as long as accessing $data[0] is safe)

``` php
        $ffi = \FFI::cdef(<<<C
            size_t zend_get_page_size(void);
        C);

        $pagesize = $ffi->zend_get_page_size();

        /* Get the address of $data */
        $start = $ffi->cast("uintptr_t", $ffi->cast("char*", $data))->cdata;
        /* Accessing memory in the same page as $start is safe */
        $max = min(self::MAX_STRING_LENGTH, ($start | ($pagesize - 1)) - $start);

        for ($i = 0; $i < $max; ++$i) {
            $result[$i] = $data[$i];

            if ("\0" === $data[$i]) {
                return implode('', $result);
            }
        }
```

File: src/Symfony/Component/VarDumper/Tests/Caster/FFICasterTest.php [hunk 3]
```diff
@@ -191,25 +190,13 @@ public function testCastCuttedPointerToChar()
...
-     *
-     * @see FFICaster::castFFIStringValue()
-     */
     public function testCastNonTrailingCharPointer()
     {
         $actualMessage = 'Hello World!';
         $actualLength = \strlen($actualMessage);
 
-        $string = \FFI::cdef()->new('char['.$actualLength.']');
+        $string = \FFI::cdef()->new('char['.($actualLength + 1).']');
```
Comment (arnaud-lb) [inline src/Symfony/Component/VarDumper/Tests/Caster/FFICasterTest.php:198]: :+1:

File: src/Symfony/Component/VarDumper/Caster/FFICaster.php (hunk 2 above)
Comment (alexandre-daubois) [reply src/Symfony/Component/VarDumper/Caster/FFICaster.php]: I couldn't find a proper way to get the page size. This looks good, thanks Arnaud!

**Review by arnaud-lb — APPROVED** [review]
Looks good to me!

**Comment by fabpot** [comment]
Thank you @alexandre-daubois.

**Pull request merged**